### PR TITLE
Changed scheduler to use deques instead of lists

### PR DESCRIPTION
**Pull request by NadavShmayo**
Currently the scheduler uses lists to store the running, waiting and swapped requests.
When iterating over each state queue we pop the first item and append to a new list, which is not good for performance since each pop is O(N) time complexity, which means the scheduler currently runs in O(N^2) time complexity (a pop happens for each item in the state queue).
Instead of using lists we could use deques, since we can pop the first item from a deque in O(1) time complexity, making the scheduler run in O(N) time complexity instead of O(N^2).

I wasn't sure whether I should keep the same types in the SchedulerOutputs class and cast from a deque to a list before returning the result, but since it seems to work with the SchedulerOutputs contains deques instead of lists I decided to change the types.

### CODE REVIEW

File: vllm/core/policy.py [hunk 1]
```diff
@@ -15,13 +15,14 @@ def get_priority(
...
         now: float,
-        seq_groups: List[SequenceGroup],
-    ) -> List[SequenceGroup]:
-        return sorted(
-            seq_groups,
-            key=lambda seq_group: self.get_priority(now, seq_group),
-            reverse=True,
-        )
+        seq_groups: deque[SequenceGroup],
+    ) -> deque[SequenceGroup]:
```
Comment (WoosukKwon) [inline vllm/core/policy.py:19]: nit: Could you use `Deque` in `typing` package for type annotation? Same for other places.
```python
from typing import Deque
```

File: vllm/core/scheduler.py [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import enum
 import time
 from typing import Dict, Iterable, List, Optional, Tuple, Union
+from collections import deque
```
Comment (WoosukKwon) [inline vllm/core/scheduler.py:4]:
```suggestion
from collections import deque
import enum
import time
from typing import Dict, Iterable, List, Optional, Tuple, Union
```
style nit: Please alphabetically sort the import statements 😄

File: vllm/core/scheduler.py [hunk 3]
```diff
@@ -125,13 +125,13 @@ def _schedule(self) -> SchedulerOutputs:
 
         # Join waiting sequences if possible.
         if not self.swapped:
-            ignored_seq_groups: List[SequenceGroup] = []
-            scheduled: List[SequenceGroup] = []
+            ignored_seq_groups: deque[SequenceGroup] = deque()
+            scheduled: deque[SequenceGroup] = deque()
```
Comment (WoosukKwon) [inline vllm/core/scheduler.py:129]: For `ignored_seq_groups` and `scheduled`, we only append new items to them. Why don't we keep using `List` for them?

File: vllm/core/scheduler.py [hunk 4]
```diff
@@ -125,13 +125,13 @@ def _schedule(self) -> SchedulerOutputs:
...
             # The total number of sequences on the fly, including the
             # requests in the generation phase.
             num_curr_seqs = sum(seq_group.get_max_num_running_seqs()
                                 for seq_group in self.running)
-            seq_lens: List[int] = []
+            seq_lens: deque[int] = deque()
```
Comment (WoosukKwon) [inline vllm/core/scheduler.py:134]: ditto. Let's keep using `List` here.

File: vllm/core/scheduler.py [hunk 5]
```diff
@@ -214,14 +217,14 @@ def _schedule(self) -> SchedulerOutputs:
         self.running = self.policy.sort_by_priority(now, self.running)
 
         # Reserve new token slots for the running sequence groups.
-        running: List[SequenceGroup] = []
-        preempted: List[SequenceGroup] = []
+        running: deque[SequenceGroup] = deque()
+        preempted: deque[SequenceGroup] = deque()
```
Comment (WoosukKwon) [inline vllm/core/scheduler.py:221]: ditto. Let's keep using `List` here.

File: vllm/core/scheduler.py [hunk 6]
```diff
@@ -275,7 +278,7 @@ def _schedule(self) -> SchedulerOutputs:
             blocks_to_swap_in=blocks_to_swap_in,
             blocks_to_swap_out=blocks_to_swap_out,
             blocks_to_copy=blocks_to_copy,
-            ignored_seq_groups=[],
+            ignored_seq_groups=deque(),
```
Comment (WoosukKwon) [inline vllm/core/scheduler.py:281]:
```suggestion
            ignored_seq_groups=[],
```

File: vllm/core/scheduler.py [hunk 7]
```diff
@@ -166,12 +166,14 @@ def _schedule(self) -> SchedulerOutputs:
...
+                    self.waiting.popleft()
                     continue
 
                 # If the number of batched tokens exceeds the limit, stop.
-                new_seq_lens = seq_lens + [num_prompt_tokens]
-                num_batched_tokens = len(new_seq_lens) * max(new_seq_lens)
+                max_seq_len = max(
+                    max(seq_lens),
+                    num_prompt_tokens) if seq_lens else num_prompt_tokens
+                num_batched_tokens = (len(seq_lens) + 1) * max_seq_len
```
Comment (WoosukKwon) [inline vllm/core/scheduler.py:176]: Why do we need this change?

File: vllm/core/scheduler.py [hunk 8]
```diff
@@ -183,12 +185,13 @@ def _schedule(self) -> SchedulerOutputs:
                         self.scheduler_config.max_num_seqs):
                     break
 
-                num_paddings = num_batched_tokens - sum(new_seq_lens)
+                num_paddings = num_batched_tokens - (sum(seq_lens) +
+                                                     num_prompt_tokens)
```
Comment (WoosukKwon) [inline vllm/core/scheduler.py:189]: Why do we need this change?

File: vllm/core/scheduler.py [hunk 9]
```diff
@@ -183,12 +185,13 @@ def _schedule(self) -> SchedulerOutputs:
...
                 if num_paddings > self.scheduler_config.max_paddings:
                     break
-                seq_lens = new_seq_lens
+                seq_lens.append(num_prompt_tokens)
```
Comment (WoosukKwon) [inline vllm/core/scheduler.py:192]: Why do we need this change?

**Review by WoosukKwon — CHANGES_REQUESTED** [review]
Hi @NadavShmayo, thanks for submitting the PR! Yes, it seems we should use deque instead of list for the queues in our scheduler. I left some minor comments on the PR. Please take a look at them!

**Review by WoosukKwon — APPROVED** [review]
LGTM. I've made some minor changes to accelerate the merge. Thanks again for the PR!

**Comment by NadavShmayo** [comment]
> LGTM. I've made some minor changes to accelerate the merge. Thanks again for the PR!

Hey, sorry for the delayed response.
I didn't get to fixing the code yet, meant to do it tomorrow, but good to see you already did!

Regarding the changes in the `new_seq_lens` logic, I thought it was better this way, instead of creating a new list each time to calculate the total batched tokens which might be slow.
But these are minor changes anyways.

Thanks for the review! 😄

**Pull request merged**